Add databaseNameXorId type

mhxbe opened the pull request:
Add databaseNameXorId type  which forces the usage of at least databaseName or databaseId but not both. Apply this to 10 impacted functions.

---

This PR resolves #232
@j-berman pointed me in the right direction.

Unfortunately, I'm running into errors when running `npm run test:types` in `src/userbase-js`
- Node: v12.18.3
- NPM: 6.14.8
```
$ npm run test:types

> userbase-js@2.3.0 test:types /Users/mikey/Desktop/playground/userbase/src/userbase-js
> dtslint types

Error: Errors in typescript@2.8 for external dependencies:
../node_modules/@types/parsimmon/index.d.ts(394,35): error TS2370: A rest parameter must be of an array type.

    at /Users/mikey/Desktop/playground/userbase/src/userbase-js/node_modules/dtslint/bin/index.js:193:19
    at Generator.next (<anonymous>)
    at fulfilled (/Users/mikey/Desktop/playground/userbase/src/userbase-js/node_modules/dtslint/bin/index.js:6:58)
```

Any idea of working around this? I would like to add tests.

j-berman commented:
Was getting the same error -- I fixed by just updating `dtslint` and `typescript` to the latest version in this commit: https://github.com/smallbets/userbase/commit/d0f5e9f321afeb00883f2a6560ecafa38ffab4ea

Pulling the latest from master and running npm install again should work

j-berman commented on src/userbase-js/types/index.d.ts:
```diff
@@ -120,29 +122,29 @@ export interface Userbase {
-  openDatabase(params: { databaseName?: string, databaseId?: string, changeHandler: DatabaseChangeHandler }): Promise<void>
+  openDatabase(params: databaseNameXorId & { changeHandler: DatabaseChangeHandler }): Promise<void>
 
   getDatabases(params?: { databaseName?: string, databaseId?: string }): Promise<DatabasesResult>
```
Could be:

`params? : databaseNameXorId`

mhxbe replied:
Fixed in commit https://github.com/smallbets/userbase/pull/236/commits/c221285795975a877eaddf6868caa058947ee191

mhxbe commented:
@j-berman, once I merged the upstream's master, the tests would run again. Thanks!

Update:
- I've added the missing **databaseNameXorId** on `getDatabases()` that you've spotted.
- For each impacted function, added $ExpectError where neather **databaseName** or **databaseId** is passed
- Add tests for missing `deleteItem()`

I saw you've also added **userbase-sql.js** where these files are also present. (index.d.ts, ...)
Ideally, these changes should also be reflected there.

Should we fix that in this PR or will you handle it in another PR?

j-berman commented:
@mhxbe I'll handle in another PR, thank you!!
